Add ability to set preview time in editor

cdwcgt opened the pull request:
close #14213
I'm not sure if having it directly in editorBeatmap as a Bindable is the best way but it seems good

![image](https://user-images.githubusercontent.com/34775378/207912532-2bb9ae48-dc60-472c-a64c-1c7a8ee4decb.png)

the green one is preview timing point(yellow already occupied by a non-inheritance point)

cdwcgt committed:
Add ability to set preview time

cdwcgt committed:
Add test for set preview point

bdach commented on osu.Game/Screens/Edit/Editor.cs:
```diff
@@ -322,6 +322,13 @@ private void load(OsuConfigManager config)
+                                    {
+                                        Items = new MenuItem[]
+                                        {
+                                            new EditorMenuItem("Set Current Position as Preview Point", MenuItemType.Standard, SetCurrectTimeAsPreview)
```
none of the other menu items are cased like this. i also don't know why this is talking about "position"

```suggestion
                                            new EditorMenuItem("Set current time as preview point", MenuItemType.Standard, SetCurrectTimeAsPreview)
```

Joehuu replied:
FWIW that is stable's wording/casing, but I agree with the suggestion above.

cdwcgt replied:
Maybe this needs to be discussed? Because I'm doing this to correspond with stable

bdach commented on osu.Game/Screens/Edit/EditorBeatmap.cs:
```diff
@@ -86,6 +86,8 @@ public partial class EditorBeatmap : TransactionalCommitComponent, IBeatmap, IBe
         [Resolved]
         private EditorClock editorClock { get; set; }
 
+        public BindableInt PreviewTime;
```
should not be publically settable like this

bdach commented on osu.Game/Screens/Edit/Editor.cs:
```diff
@@ -801,6 +808,11 @@ protected void Clone()
 
         protected void Redo() => changeHandler?.RestoreState(1);
 
+        protected void SetCurrectTimeAsPreview()
```
typo in method name

bdach commented on osu.Game/Screens/Edit/EditorBeatmap.cs:
```diff
@@ -107,6 +109,9 @@ public EditorBeatmap(IBeatmap playableBeatmap, ISkin beatmapSkin = null, Beatmap
             foreach (var obj in HitObjects)
                 trackStartTime(obj);
+
+            PreviewTime = new BindableInt(playableBeatmap.Metadata.PreviewTime);
```
isn't handling -1 correctly

https://github.com/ppy/osu/blob/60c8ef3fe5fe9adcb204814d975b7ed6c832fc8c/osu.Game/Beatmaps/BeatmapMetadata.cs#L48-L51

bdach commented on osu.Game/Screens/Edit/Components/Timelines/Summary/Parts/PreviewTimePart.cs:
```diff
@@ -0,0 +1,31 @@
+            }
+
+            [BackgroundDependencyLoader]
+            private void load(OsuColour colours) => Colour = colours.Lime;
```
lime is already being used by difficulty points

https://github.com/ppy/osu/blob/60c8ef3fe5fe9adcb204814d975b7ed6c832fc8c/osu.Game/Beatmaps/ControlPoints/DifficultyControlPoint.cs#L31

cdwcgt replied:
I'm not sure which color to use, I need advice

LeNitrous replied:
Stable uses yellow
![image](https://user-images.githubusercontent.com/20495991/210075519-4c10086c-50dd-4b2f-9434-8c4f69da0de6.png)

bdach replied:
not sure yellow is going to work either as lazer uses that for timing points

cdwcgt committed:
only get;

cdwcgt committed:
`PreviewTimePart` will not show if preview time is -1

cdwcgt committed:
fix typo

bdach commented on osu.Game.Tests/Visual/Editing/TestScenePreviewTime.cs:
```diff
@@ -0,0 +1,35 @@
+        public void TestScenePreviewTimeline()
+        {
+            AddStep("set preview time to -1", () => EditorBeatmap.PreviewTime.Value = -1);
+            AddAssert("preview time line should not show", () => Editor.ChildrenOfType<PreviewTimePart>().Single().Alpha == 0);
```
why shouldn't it show? if `PreviewTime` is -1 then it's automatically 40% of the track length. surely you can show the line there

cdwcgt replied:
This should be the expected situation.
The preview point should not be displayed on the map without preview point (also stable)

If it is really shown in the 40% position, it may be misleading (let mapper think its already set)
if we must do this, I think it should be a different color, but that would get weird

peppy replied:
I think of these options, not showing at all is probably best.

bdach commented on osu.Game/Screens/Edit/Components/Timelines/Summary/Parts/PreviewTimePart.cs:
```diff
@@ -0,0 +1,35 @@
+        protected override void LoadBeatmap(EditorBeatmap beatmap)
+        {
+            base.LoadBeatmap(beatmap);
+            Add(new PreviewTimeVisualisation(beatmap.PreviewTime));
```
also just noticed that this is incorrect, you cannot pass bindables like this to component - make a local bindable in `PreviewTimeVisualisation` and bind to `beatmap.PreviewTime` instead

cdwcgt committed:
safe way to pass bindable

cdwcgt committed:
readonly

bdach committed:
Merge branch 'master' into previewTime

bdach committed:
Adjust method naming and copy

bdach committed:
Add all possible timeline elements to test for demonstrative purposes

bdach committed:
Adjust visual appearance of preview time part

- Use slightly different hue of green to distinguish from difficulty
  control points. The colour is still not ideal, but picking
  a distinctive enough hue is pretty hard.

- Place the preview time part at the bottom rather at the top.
  Not sure why it was at the top; not only could it overlap with the
  control points, but it also looked quite badly misaligned there
  when bookmarks were displayed at the bottom.

bdach committed:
Adjust binding logic

bdach committed:
Fix changing preview point not prompting for save

bdach commented:
I've taken over this PR to hopefully bring it to a conclusion. Of note, while there are objectively correct fixes there (https://github.com/ppy/osu/pull/21650/commits/b689ad6d80e177734a470fc0d30e6497dc9af09b), I've also made subjective adjustments to the copy of the menu item to be more in line with everything else around it (can't care less about what stable's specific wording was), and also changed a few things in the visual appearance of the preview point marker in timeline.

Before:

![image](https://user-images.githubusercontent.com/20418176/210181836-f15d9fa3-8f22-40bd-9734-44ddae000519.png)

After:

![image](https://user-images.githubusercontent.com/20418176/210181862-feb97fc9-fd4a-4ae8-8ae5-893356dd7a94.png)

In particular:

- Marker uses slightly different hue of green to distinguish it from difficulty control points
- Marker no longer overlaps control point markers and is shown in the same manner as bookmarks are

Dunno if people will agree but these feel pretty no-brainer to me. Will need a second review.

frenzibyte committed:
Use sane `BeatmapInfo` for preview time mutation

`EditorBeatmap.BeatmapInfo` is usually the correct instance for mutating
properties that should persist in the database.

frenzibyte committed:
Rewrite visualisation piece to bind once and without potential event leak

frenzibyte reviewed (approved):
This looks pretty fine to me. ~~I was hoping we can share the preview time at a `BeatmapSetInfo` level since we only allow one track per set, but that didn't turn out to be an easy change so I'm leaving it for later.~~

bdach commented:
> I was hoping we can share the preview time at a `BeatmapSetInfo` level since we only allow one track per set

why would that even be in the scope of this change though. that is a wild swing of a thing to attempt without even trying to communicate prior intent? was this talked about anywhere?

frenzibyte commented:
Considered it a small UX change that would definitely be nice to have rather than forcing the mapper to edit each difficulty to set the same preview point.

bdach commented:
> Considered it a small UX change that would definitely be nice to have rather than forcing the mapper to edit each difficulty to set the same preview point

this is something to revisit with something like https://github.com/ppy/osu/issues/18903.

also i don't think stable enforces this and i don't know whether mappers don't want the freedom of setting preview point for whatever reason.

> (and it should be just using `BeatmapSetInfo.Metadata` in theory, but something with caching is breaking it from fully working).

that sounds incorrect because if this is to become a set-level operation then i'm pretty sure you need to re-encode all `.osu` files for all difficulties in the set to be fully correct operation. but again why are we even talking about this yet _without prior discussion_.

frenzibyte commented:
Okay, let's just ignore it.

frenzibyte committed:
Merge branch 'master' into previewTime

frenzibyte committed:
Merge branch 'master' into previewTime
